src/util/piped_process.cpp

```diff
@@ -0,0 +1,168 @@
+/// \file Subprocess communication with pipes
+///
+/// \author Diffblue Ltd.
+
+#ifdef _WIN32
+// Windows includes go here
+#else
+#  include <cstring>
+#  include <errno.h>
+#  include <fcntl.h> // library for fcntl function
+#  include <iostream>
+#  include <unistd.h>
+#endif
+
+#include "invariant.h"
+#include "piped_process.h"
+
+piped_processt::piped_processt(const std::string &command)
+{
+#ifdef _WIN32
+  // This should use the new error state from PR #6131 once that is done
+  INVARIANT(false, "New SMT2 backend WIP: Windows piped_process constructor.");
+#else
+
+  if(pipe(pipe_input) == -1)
+  {
+    throw std::runtime_error("Input pipe creation failed");
+  }
+
+  if(pipe(pipe_output) == -1)
+  {
+    throw std::runtime_error("Output pipe creation failed");
+  }
+
+  // Default state
+  process_state = process_statet::NOT_CREATED;
+
+  if(fcntl(pipe_output[0], F_SETFL, O_NONBLOCK) < 0)
+  {
+    throw std::runtime_error("Setting pipe non-blocking failed");
+  }
+
+  // Create a new process for the child that will execute the
+  // command and receive information via pipes.
+  pid_t pid = fork();
+  if(pid == 0)
+  {
+    // child process here
+
+    // Close pipes that will be used by the parent so we do
+    // not have our own copies and conflicts.
+    close(pipe_input[1]);
+    close(pipe_output[0]);
+
+    // Duplicate pipes so we have the ones we need.
+    dup2(pipe_input[0], STDIN_FILENO);
+    dup2(pipe_output[1], STDOUT_FILENO);
+    dup2(pipe_output[1], STDERR_FILENO);
+
+    // Create the arguments to execvp from the construction string
+    char **args = split_command_args(command);
```

Comment:
:question: Where does this raw pointer get freed up?

Reply:
There are two cases for this:

1) If `execvp` succeeds, it's not handled anywhere - and it doesn't need to, for the simple reason that the process image changes to the binary invoked in `exec` so the pointer is no longer present there.
2) In case `execvp` fails, it's freed later on, but there could be an argument made that it wouldn't need to be handled there as well, as the failure of execvp is leading to a terminal exception.